### PR TITLE
Python API for restoring delta table

### PR DESCRIPTION
**Pull request by Maks-D**
* Add possibility to restore delta table using version or timestamp from pyspark
   Examples:
   ```
   DeltaTable.forPath(spark, path).restoreToVersion(0)
   DeltaTable.forPath(spark, path).restoreToTimestamp('2021-01-01 01:01-01')
   ```

Tested by unit tests.

Fixes https://github.com/delta-io/delta/issues/890

Signed-off-by: Maksym Dovhal <maksym.dovhal@gmail.com>

### CODE REVIEW

**Comment by Maks-D** [comment]
Hi, @tdas,
Just want to inform you that I've removed cache of filesToRemove from RestoreDeltaTable (according to our discussion https://github.com/delta-io/delta/pull/863#issuecomment-1015672532)

File: python/delta/tests/test_deltatable.py [hunk 1]
```diff
@@ -791,6 +791,34 @@ def test_protocolUpgrade(self) -> None:
...
+        self.__writeDeltaTable([('a', 1), ('b', 2)])
+        timestampToRestore = DeltaTable.forPath(self.spark, self.tempFile) \
+            .history() \
+            .head() \
+            .timestamp \
+            .strftime('%Y-%m-%d %H:%M:%S.%f')
+
+        self.__overwriteDeltaTable([('a', 3), ('b', 2)],
+                                   schema=["key_new", "value_new"],
+                                   overwriteSchema='true')
```
Comment (vkorukanti) [inline python/delta/tests/test_deltatable.py:814]: sorry its not clean on why do we need to specify the new schema and overwrite it. Isn't it the same schema as the v0?

File: python/delta/tests/test_deltatable.py [hunk 2]
```diff
@@ -791,6 +791,34 @@ def test_protocolUpgrade(self) -> None:
...
+
```
Comment (vkorukanti) [inline python/delta/tests/test_deltatable.py:815]: Add check to verify that data has changed?
`self.__checkAnswer(restored, [Row(key='a', value=3), Row(key='b', value=2)])`?

Similarly in the above test

**Review by vkorukanti — CHANGES_REQUESTED** [review]
Thanks @Maks-D for the PR. The PR looks good. Minor comments. I am wondering if you could split this into two PRs (one PR handling the python API and other other one handling the refactoring)

File: python/delta/tests/test_deltatable.py (hunk 1 above)
Comment (Maks-D) [reply python/delta/tests/test_deltatable.py:814]: `__overwriteDeltaTable` overwrites table with a new schema and new data. This helps us to show that restore% methods restores schema and data.
Some code examples to show this in details:
```python
df = spark.createDataFrame([(1, 'a')], ["key", "value"])
df.write.format("delta").mode("overwrite").save("/tmp/delta_test")
spark.read.format("delta").load("/tmp/delta_test").printSchema()
root
 |-- key: long (nullable = true)
 |-- value: string (nullable = true)

df = spark.createDataFrame([(1, 'a')], ["key_new", "value_new"])
df.write.format("delta").mode("overwrite").option('overwriteSchema', 'true').save("/tmp/delta_test")
spark.read.format("delta").load("/tmp/delta_test").printSchema()
root
 |-- key_new: long (nullable = true)
 |-- value_new: string (nullable = true)
```

Comment (vkorukanti) [reply python/delta/tests/test_deltatable.py:818]: Ah.. I see. that makes sense. Thanks for the explanation.

**Comment by Maks-D** [comment]
@vkorukanti Thank you for review.
I've fixed your comments. Refactoring moved to https://github.com/delta-io/delta/pull/912

File: python/delta/tests/test_deltatable.py (hunk 2 above)
Comment (Maks-D) [reply python/delta/tests/test_deltatable.py:820]: Both tests are updated

**Pull request closed without merging**